http/src/main/scala/org/http4s/blaze/http/ClientSessionManagerImpl.scala

```diff
@@ -0,0 +1,167 @@
+    override def close(within: Duration): Future[Unit] = parent.close(within)
+  }
+
+  def apply(config: HttpClientConfig): ClientSessionManagerImpl = {
+    val sessionCache = new java.util.HashMap[ConnectionId, java.util.Collection[HttpClientSession]]
+    new ClientSessionManagerImpl(sessionCache, config)
+  }
+}
+
+private final class ClientSessionManagerImpl(sessionCache: java.util.Map[ConnectionId, java.util.Collection[HttpClientSession]], config: HttpClientConfig) extends ClientSessionManager {
+
+  private[this] def getId(composition: UrlComposition): ConnectionId =
+    ConnectionId(composition.scheme, composition.authority)
+
+  private[this] implicit def ec = Execution.trampoline
+
+  private[this] val logger = getLogger
+  private[this] val socketFactory = new ClientChannelFactory(group = config.channelGroup)
+
+  override def acquireSession(request: HttpRequest): Future[HttpClientSession] = {
+    logger.debug(s"Acquiring session for request $request")
+     UrlComposition(request.url) match {
+      case Success(urlComposition) =>
+        val id = getId(urlComposition)
+        val session = findExistingSession(id)
+
+        if (session == null) createNewSession(urlComposition, id)
+        else {
+          logger.debug(s"Found hot session for id $id: $session")
+          Future.successful(session)
+        }
+
+      case Failure(t) => Future.failed(t)
+    }
+  }
+
+  // WARNING: can emit `null`. For internal use only
+  private[this] def findExistingSession(id: ConnectionId): HttpClientSession = sessionCache.synchronized {
+    sessionCache.get(id) match {
+      case null => null // nop
+      case sessions =>
+        var session: HttpClientSession = null
+        val it = sessions.iterator
+        while(session == null && it.hasNext) it.next() match {
+          case h2: Http2ClientSession if h2.status == Closed =>
+            h2.closeNow() // make sure its closed
+            it.remove()
+
+          case h2: Http2ClientSession if h2.quality > 0.1 =>
+            session = h2
+
+          case _: Http2ClientSession => () // nop
+
+          case h1: Http1ClientSession =>
+            it.remove()
+            if (h1.status != Closed) { // Should never be busy
+              session = h1 // found a suitable HTTP/1.x session.
+            }
+        }
+
+        // if we took the last session, drop the collection
+        if (sessions.isEmpty) {
+          sessionCache.remove(id)
+        }
+
+        session
+    }
+  }
+
+  private[this] def createNewSession(urlComposition: UrlComposition, id: ConnectionId): Future[HttpClientSession] = {
+    logger.debug(s"Creating new session for id $id")
+    val p = Promise[HttpClientSession]
+
+    socketFactory.connect(urlComposition.getAddress).onComplete {
+      case Failure(e) => p.tryFailure(e)
+      case Success(head) =>
+        val clientStage = new Http1ClientStage(config)
+        var builder = LeafBuilder(clientStage)
+        if (urlComposition.scheme == "https") {
```

Comment:
equalsIgnoreCase, or is this forced lowercase elsewhere?

Reply:
Nope, not that I recall, and it's better to do so to prevent regressions anyway. I'll change it.